### PR TITLE
UD wallet tokens support

### PR DESCRIPTION
**Pull request by supermassive**
Resolves https://github.com/brave/brave-browser/issues/25562

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
Wallet 'Send to' supports resolving different tokens(if setup by domain owner)
For example for jim-unstoppable.crypto
Token `UNI` for chains `Ethereum mainnet`, `Binance Smart Chain Mainnet` and `Harmony Mainnet Shard 0` resolves to different addresses.
`SOL` token for Solana networks is also resolvable to solana-kind addresses
`FIL` is also supported(try https://ud.me/uns-devtest-brave.x)

### CODE REVIEW

File: components/brave_wallet/browser/brave_wallet_constants.h [hunk 1]
```diff
@@ -976,6 +976,8 @@ constexpr char kPolygonMainnetEndpoint[] = "https://mainnet-polygon.brave.com/";
 
 // Unstoppable domains record key for ethereum address.
 constexpr char kCryptoEthAddressKey[] = "crypto.ETH.address";
+// Unstoppable domains record key for solana address.
+constexpr char kCryptoSolAddressKey[] = "crypto.SOL.address";
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/brave_wallet_constants.h:980]: https://github.com/brave/brave-core/pull/15388/commits/bc87e3d381ba6d140e57d27dbc20b0a28d6dd269

File: components/brave_wallet/browser/eth_data_builder.cc [hunk 2]
```diff
@@ -23,6 +28,25 @@ bool IsValidHostLabelCharacter(char c, bool is_first_char) {
...
+      {"0x1", "ERC20"},
+      {"0x38", "BEP20"},
+      {"0x63564c40", "HRC20"},
+      {"0x64", "XDAI"},
+      {"0x7a", "FUSE"},
+      {"0x89", "MATIC"},
+      {"0xa", "OP"},
+      {"0xa4b1", "AETH"},
+      {"0xa86a", "AVAX"},
+      {"0xfa", "FANTOM"},
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/eth_data_builder.cc:45]: added corner case support for crypto.FTM.version.OPERA.address
https://github.com/brave/brave-core/pull/15388/commits/bc87e3d381ba6d140e57d27dbc20b0a28d6dd269

File: components/brave_wallet/browser/eth_data_builder.cc [hunk 3]
```diff
@@ -284,29 +308,61 @@ absl::optional<std::string> GetMany(const std::vector<std::string>& keys,
...
 
-  return data;
+  return lookup_keys;
+}
+
+std::vector<uint8_t> GetWalletAddr(const std::string& domain,
+                                   mojom::CoinType coin,
+                                   const std::string& symbol,
+                                   const std::string& chain_id) {
+  DCHECK(coin == mojom::CoinType::ETH || coin == mojom::CoinType::SOL);
```
[the remark this answers is not in the document]
Comment (supermassive) [reply components/brave_wallet/browser/eth_data_builder.cc]: https://github.com/brave/brave-core/commit/bc87e3d381ba6d140e57d27dbc20b0a28d6dd269

**Review by yrliou — APPROVED** [review]
backend++, please list the exact address we expect for different tokens linked by jim-unstoppable.crypto in the test plan so it would be easier for QA to follow and check.

File: components/brave_wallet/browser/eth_data_builder.cc [hunk 4]
```diff
@@ -23,6 +30,32 @@ bool IsValidHostLabelCharacter(char c, bool is_first_char) {
          (c >= '0' && c <= '9') || (!is_first_char && c == '-') || c == '_';
 }
 
+absl::optional<std::string> ChainIdToVersion(const std::string& symbol,
```
Comment (onyb) [inline components/brave_wallet/browser/eth_data_builder.cc:33]: Should this be inside the `unstoppable_domains` namespace?

**Review by onyb — CHANGES_REQUESTED** [review]
Left a small remark

File: components/brave_wallet/browser/eth_data_builder.cc (hunk 4 above)
Comment (supermassive) [reply components/brave_wallet/browser/eth_data_builder.cc:33]: That's in anonymous namespace, only for local use

**Review by onyb — APPROVED** [review]
++

**Pull request merged**